### PR TITLE
Date/Time Types and Serializers [API-1230]

### PR DESCRIPTION
**Pull request by yuce**
This PR adds the following date/time types and their serializers.

* types.LocalTime
* types.LocalDate
* types.LocalDateTime
* types.OffsetDateTime

### CODE REVIEW

**Review by srknzl — APPROVED** [review]
Approving with some wording improvement comments

File: sql/driver/doc.go [hunk 1]
```diff
@@ -175,12 +175,28 @@ The names in parentheses correspond to SQL types:
...
 	- time.Time (time) Detected by checking: year == 0, month == day == 1
 	- time.Time (timestamp) Detected by checking: not time, timezone == time.Local
 	- time.Time (timestamp with time zone) Detected by checking: not time, timezone != time.Local
 	- serialization.JSON (json)
 
+Using Date/Time
+
+time.Time values are automatically serialized to the correct type.
+
+In order to force using a specific date/time type, create a time.Time value and cast it to the target type:
```
[the remark this answers is not in the document]
Comment (yuce) [reply sql/driver/doc.go:192]: `time.Time` is the `Time` type from `time` package.

File: internal/serialization/serialization_improvements_test.go [hunk 2]
```diff
@@ -102,23 +102,43 @@ func TestSerializationImprovements(t *testing.T) {
 		}{
 			{
 				input:  time.Date(2021, 2, 10, 0, 0, 0, 0, time.Local),
-				name:   "JavaLocalDate",
-				target: time.Date(2021, 2, 10, 0, 0, 0, 0, time.Local),
+				name:   "JavaLocalDate from time.LocalTime",
```
[the remark this answers is not in the document]
Comment (yuce) [reply internal/serialization/serialization_improvements_test.go]: That's copy/paste hazard I think. Fixed at: https://github.com/hazelcast/hazelcast-go-client/pull/741/commits/0438c2556cbc22f3ee7d4b6bdd00391cde68bddb

File: types/time.go [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+	"time"
+)
+
+// LocalDate is the date part of time.Time.
+type LocalDate time.Time
+
+// LocalTime is the time part of time.Time.
+type LocalTime time.Time
+
+// LocalDateTime is the date time with local time zone.
```
[the remark this answers is not in the document]
Comment (yuce) [reply types/time.go]: Changed to `date and time` at: https://github.com/hazelcast/hazelcast-go-client/pull/741/commits/0438c2556cbc22f3ee7d4b6bdd00391cde68bddb

File: serialization/api.go [hunk 4]
```diff
@@ -287,22 +285,22 @@ type PortableWriter interface {
...
+	WriteTime(fieldName string, t *types.LocalTime)
+	// WriteTimestamp writes the date and time without the timezone offset.
+	WriteTimestamp(fieldName string, t *types.LocalDateTime)
+	// WriteTimestampWithTimezone writes the date and time with the timezone offset.
+	WriteTimestampWithTimezone(fieldName string, t *types.OffsetDateTime)
+	// WriteDateArray writes a date array.
+	WriteDateArray(fieldName string, t []types.LocalDate)
+	// WriteTimeArray writes a time array.
+	WriteTimeArray(fieldName string, t []types.LocalTime)
+	// WriteTimestampArray writes dates and times.
```
[the remark this answers is not in the document]
Comment (yuce) [reply serialization/api.go]: Changed at: https://github.com/hazelcast/hazelcast-go-client/pull/741/commits/0438c2556cbc22f3ee7d4b6bdd00391cde68bddb

File: sql/driver/doc.go [hunk 5]
```diff
@@ -175,12 +175,28 @@ The names in parentheses correspond to SQL types:
...
+	- types.LocalTime (time)
+	- types.LocalDateTime (timestamp)
+	- types.OffsetDateTime (timestamp with time zone)
 	- time.Time (date) Detected by checking: hour == minute == second == nanoseconds = 0
 	- time.Time (time) Detected by checking: year == 0, month == day == 1
 	- time.Time (timestamp) Detected by checking: not time, timezone == time.Local
 	- time.Time (timestamp with time zone) Detected by checking: not time, timezone != time.Local
 	- serialization.JSON (json)
 
+Using Date/Time
```
[the remark this answers is not in the document]
Comment (yuce) [reply sql/driver/doc.go]: Updated at: https://github.com/hazelcast/hazelcast-go-client/pull/741/commits/0438c2556cbc22f3ee7d4b6bdd00391cde68bddb

File: sql/driver/doc.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (srknzl) [reply sql/driver/doc.go:192]: :D sorry.

File: internal/serialization/default_portable_reader.go [hunk 6]
```diff
@@ -346,66 +347,70 @@ func (pr *DefaultPortableReader) GetRawDataInput() serialization.DataInput {
...
 	})
 	return
 }
 
-func (pr *DefaultPortableReader) ReadTimestampWithTimezone(fieldName string) (t *time.Time) {
+func (pr *DefaultPortableReader) ReadTimestampWithTimezone(fieldName string) (t *types.OffsetDateTime) {
 	pr.readNullable(fieldName, serialization.TypeTimestampWithTimezone, func() {
 		v := ReadPortableTimestampWithTimezone(pr.input)
-		t = &v
+		t = (*types.OffsetDateTime)(&v)
```
Comment (utku-caglayan) [inline internal/serialization/default_portable_reader.go:377]: What about
```suggestion
		v := types.OffsetDateTime(ReadPortableTimestampWithTimezone(pr.input))
		t = &v
```

File: sql_it_test.go [hunk 7]
```diff
@@ -390,6 +439,91 @@ func TestSQLWithPortableDateTime(t *testing.T) {
 	})
 }
 
+func TestSQLWithPortableDateTime2(t *testing.T) {
```
Comment (utku-caglayan) [inline sql_it_test.go:442]: If this does this bring any extra value over the similar test on "driver/driver_it_test.go", I think we can remove one of them. Wdyt?

**Review by utku-caglayan — COMMENTED** [review]
Noted a few questions and suggestions

File: sql_it_test.go (hunk 7 above)
Comment (yuce) [reply sql_it_test.go:442]: They are mostly similar but not the same. This one tests the SQL methods on the Client, and the other one tests the driver.

File: internal/serialization/default_portable_writer.go [hunk 8]
```diff
@@ -184,48 +185,56 @@ func (pw *DefaultPortableWriter) WritePortableArray(fieldName string, portableAr
 	}
 }
 
-func (pw *DefaultPortableWriter) WriteDate(fieldName string, t *time.Time) {
+func (pw *DefaultPortableWriter) WriteDate(fieldName string, t *types.LocalDate) {
 	pw.writeNullableField(fieldName, serialization.TypeDate, t == nil, func() {
-		WritePortableDate(pw.output.ObjectDataOutput, *t)
+		tt := (*time.Time)(t)
+		WritePortableDate(pw.output.ObjectDataOutput, *tt)
```
[the remark this answers is not in the document]
Comment (utku-caglayan) [reply internal/serialization/default_portable_writer.go:191]: [If I am not mistaken, it does not.](https://github.com/yuce/hazelcast-go-client/blob/dt-wrappers/internal/serialization/default_portable_writer.go#L290) I guess the way I shared makes to copies, so we can ignore it.

**Review by utku-caglayan — APPROVED** [review]
Looks great!

**Comment by yuce** [comment]
@utku-caglayan @srknzl Thanks for the review!

**Pull request merged**